### PR TITLE
`IdentifierHiding`: Improve performance, address some false positives/false negatives

### PR DESCRIPTION
**Pull request by lcartey**
### Background
`IdentifierHiding.qll` has been frequently reported as one of the slowest performing queries in the Coding Standards suite. In principle, the query is trying to achieve something quite simple, which is the following:
```codeql
exists(UserVariable hiddenVariable, UserVariable hidingVariable, Scope parentScope, Scope childScope |
  childScope.getAnAncestor() = parentScope
  hiddenVariable = parent.getAVariable() and
  hidingVariable = childScope.getAVariable() and
  hiddenVariable.getName() = hidingVariable.getName()
)
```
In words: find a pair of variables, where one is declared in a nested scope of the other, and where they have the same name. However, performance of this simple approach suffers on large codebases with a lot of name duplication, and a lot of variables declared in the global scope, because there is no ideal join order: joining by name first is expensive, as is joining by variables in a nested scope (which is the current approach, implemented in `getOuterScopesOfVariable_candidate`).

### Updated algorithm

This PR therefore rewrites the variable hiding algorithm from scratch to avoid these performance concerns, by implementing the hiding detection using a phased approach:
 * `Phase 1`: compute for each scope the set of string `names` declared in this scope or a nested scope (`Scope::isNameDeclaredInThisOrNestedScope(string name)`, `Scope::isNameDeclaredInNestedScope(string name)`), and use these to determine, for each scope, a set of variables that are potentially hidden in a nested scope (`UserVariable Scope::getAPotentiallyHiddenVariable(string name)`).
 * `Phase 2`: compute for each scope the set of `UserVariables` that are potentially hidden by a variable declared in this or a nested scope (`UserVariable Scope::getAVariableHiddenByThisOrNestedScope(string name)`).
 * `Phase 3`: compute for each scope a candidate set of hidden/hiding variables, where the hidden variable is declared in an outer scope, and the hiding variable is declared in this or a nested scope (`Scope::hidesCandidate(UserVariable hiddenVariable, UserVariable hidingVariable, string name)`)

Each phase in this approach remains tractable, and avoids the large joins on either variables by name or by scope.

In making this change, I have incorporated the detection of identifier hiding directly into the scope hiding calculation. This reduces duplication and removes further opportunities for poor join ordering performance.

In the process of this performance change, I've fixed two FP/FN issues:
 * Variables declared in nested loops are now consistently identified. This is due to consistency improvements in the `getParentScope` calculation.
 * Variable declarations in lambda expressions now consider whether the variable they hide is in the same translation unit, reducing false positives.

### Review

I would recommend a commit-wise review:
 * Preparatory commits to simplify the refactoring:
    *  27efc319d67e44524f3eeeae10b61b85f71671c0 - remove unused Scope import
    * 0c92f2efa9b6305fae4e3e188c5eab2939f71886 - switch DCL53-CPP from `hides` to `hidesStrict` (no semantic difference for this query).
    * 90d4127a7bc8c8f3a9a09541f39aebb50f25a98d - remove `hides`, as it's no longer used.
 * Adjustments to `getParentScope` to address consistency issues:
   * 6b5021a923702d9b85bca2e923ff279b8e5fd69f - adds testing for the `getParentScope` predicate.
   * b8b41312c24fa6d9ce511a5d0543ddef08c7f6ee - fixes an issues which was causing some elements to have multiple parent scopes.
   * 1f8e2baedb543b5a4318fc87d63b797865902f18 - address parent scope issues with loop initializers, conditions and increments.
   * dbd3fe6eeb10a1db03b2599765632a80539557e7 - a later commit, but this fixes issues with the parents of catch blocks.
 * Initial performance adjustments:
   * 63a60b1855f00c96a63e16242055ab2f660782f8 - improve performance of `excludedViaNestedNamespace` by inlining it late.
   * 1bd839c34c39e3546484d666ebe8b15d0e9e0451 - implement a simplified version of the improved algorithm.
   * f25507e821aeae14df6e2b676599b57768016387 and b5ff407e07197d229664b462503916f1da268104 - some refactoring improvements.
   * 61521e0144707fe65c611988f2412d86af94e07c - refactor declaration order detection for block scope defined variables to avoid requiring transitive closure on the parent scope.
 * Integrate lambda expression hiding in the scope hiding calculation:
   * 47b3fb272f27f2fd7f06ab09d0c1195a7d06a16d - modify the hiding algorithm to expose pairs of variables in `hideCandidate`. This enables as-we-go filtering, which is required to apply the lambda hiding rules.
   * 9b7e1299fc91d9845f5e9250d33102b6fb627755 - implement a `LambdaScope` class to implement the lambda hiding rules in `hidesCandidate`. This is not used yet in this commit, because lambda expressions are not tied into the `getParentScope` hierarchy.
   * 117d0fba033e58ac03d7903ada230e2a09114972 - capture lambda expressions in the `getParentScope` hierarchy.
   * 411ecde2794bb92bb6981ee6107044d66d5de4ac - remove the special casing of lambda expressions in `IdentifierHiding.qll`.

### Performance testing

I used https://github.com/grpc/grpc for initial testing and development as it contains both a large number of variables with the same name and a large number of variables declared at the global scope. On my machine, before this PR, it took around 750 seconds to evaluate `IdentifierHiding.ql` from a cold cache, with the following slow predicates:

time  | evals |   max @ iter | predicate
------|-------|--------------|----------
3m30s |       |              | Scope::excludedViaNestedNamespaces/2#117d2dae@a8378djc
2m9s |       |              | _Scope::Scope.getADeclaration/0#dispred#900017c1#fb_10#join_rhs__Scope::UserVariable#715743a1_Variab__#shared@cfddd5k1
2m2s |       |              | _Scope::UserVariable#715743a1_Variable::Variable.getName/0#dispred#1b4a78c5_Variable::Variable.getNa__#shared@57fbdf1e
59.3s |       |              | _#Scope::Scope.getStrictParent/0#dispred#23e5f1eaPlus#sinkBound#4#2#3#2#4_#Scope::Scope.getStrictPar__#higher_order_body@893f42kb
58.5s |       |              | _Scope::Scope.getADeclaration/0#dispred#900017c1#fb_10#join_rhs__Scope::UserVariable#715743a1_Variab__#shared#1@69d9cfa6
27.4s |       |              | _#Scope::Scope.getStrictParent/0#dispred#23e5f1eaPlus#sourceBound#4#3_Declaration::Declaration.hasSp__#higher_order_body@5f38a95e

After this PR, it takes around 45 seconds from a cold cache, with the following most expensive predicates:

time  | evals |   max @ iter | predicate
------|-------|--------------|----------
7.8s |    27 |  3.7s @ 11   | Scope::Scope.getAVariableHiddenByThisOrNestedScope/1#f46b1a52@d5522xlq
6.2s |     2 |  3.8s @ 2    | Element::ElementBase.toString/0#dispred#6e016f86@277e1cso
901ms |       |              | Element::Element.getFile/0#dispred#536cb5f3@f25789it

I've also run against the Top 10 C/C++ codebases on MRVA, and seen a similar performance improvement.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [ ] Queries have been added for the following rules:
     - _rule number here_
 - [x] Queries have been modified for the following rules:
     - `A2-10-1`
     - `RULE-5-3`
     - `DCL53-CPP`

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [x] Yes
  - [ ] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [x] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Review by knewbury01 — CHANGES_REQUESTED** [review]
@lcartey just one comment for now, still reviewing

File: cpp/common/src/codingstandards/cpp/Scope.qll [hunk 1]
```diff
@@ -161,21 +294,17 @@ predicate inSameTranslationUnit(File f1, File f2) {
 }
 
 /**
- * Gets a user variable which occurs in the "potential scope" of variable `v`.
- */
-cached
-UserVariable getPotentialScopeOfVariable(UserVariable v) {
```
[the remark this answers is not in the document]
Comment (lcartey) [reply cpp/common/src/codingstandards/cpp/Scope.qll:167]: Thanks! I've updated the that query to use a custom predicate to achieve the same semantic goal.

File: change_notes/2024-12-08-identifier-hiding.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+ - `A2-10-1` - `IdentifierHiding.ql`:
+   - Improved evaluation performance.
+   - Addressed false negatives where nested loops used the same variable name.
+   - Exclude cases where a variable declared in a lambda expression shadowed a globa or namespace variable that did not appear in the same translation unit.
```
Comment (fjatWbyT) [inline change_notes/2024-12-08-identifier-hiding.md:4]:
```suggestion
   - Exclude cases where a variable declared in a lambda expression shadowed a global or namespace variable that did not appear in the same translation unit.
```

File: change_notes/2024-12-08-identifier-hiding.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+ - `RULE-5-3` - `IdentifierHidingC.ql`:
```
Comment (fjatWbyT) [inline change_notes/2024-12-08-identifier-hiding.md:5]:
```suggestion
 - `RULE-5-3` - `IdentifierHiding.ql`:
```

**Review by knewbury01 — APPROVED** [review]
lgtm! thanks for all this work

File: change_notes/2024-12-08-identifier-hiding.md (hunk 3 above)
Comment (fjatWbyT) [reply change_notes/2024-12-08-identifier-hiding.md:5]: Oh, yeah, that's right, my mistake.
Thanks a lot for clarifying and my excuses for the noise 🙇🏻

Comment (knewbury01) [reply change_notes/2024-12-08-identifier-hiding.md:5]: no problem, thanks for the review! :)

**Pull request merged**